k2/csrc/fsa.h

```diff
@@ -77,13 +77,40 @@ struct Fsa {
   // contains the first arc-index leaving this state (index into `arcs`).
   // The next element of this array gives the end of that range.
   // Note: the final-state is numbered last, and implicitly has no
-  // arcs leaving it.
+  // arcs leaving it. For non-empty FSA, we put a duplicate of the final state
+  // at the end of `arc_indexes` to avoid boundary check for some FSA
+  // operations. Caution: users should never call `arc_indexes.size()` to get
+  // the number of states, they should call `NumStates()` to get the number.
   std::vector<int32_t> arc_indexes;
 
   // Note: an index into the `arcs` array is called an arc-index.
   std::vector<Arc> arcs;
 
-  StateId NumStates() const { return static_cast<StateId>(arc_indexes.size()); }
+  Fsa() {}
+  // just for creating testing FSA examples for now.
+  Fsa(std::vector<Arc> fsa_arcs, int32_t final_state) {
```

Comment:
Better to use
```cpp
Fsa(std::vector<Arc> &&fsa_arcs, int32_t final_state);
```

The user should invoke the constructor by calling `std::move` explicitly.

Reply:
just want to accept both lvalue and rvalue here.

Reply:
For lvalue, there is a copy.

Reply:
yes, for lvaue it is a copy as we must copy for lvalue.